doc/ir-semantics.text

```diff
@@ -111,18 +111,15 @@ Update (partial) function, returns an H:
 
 ## Osr transformation of environment
 
-  materialize P H E (mut x = e) = (x ↦ (a ↦ v))
-    where a - fresh
-      and v := eval P H E e
-  materialize P H E _           = ()
-
-  osr-env P H E M (const x = e) = (x ↦ val v)
+  create-env H E (const x = e) = (∅, [x ↦ val v])
     where v := eval P H E e
-  osr-env P H E M (mut x)       = (x ↦ ⊥)
-  osr-env P H E M (mut x = &y)  = (x ↦ ref a)
+  create-env H E (mut x)       = ([a ↦ ⊥], [x ↦ ref a])
+    where a fresh
```

Comment:
But that semantics means that assignment may allocate, right? (Doesn't that mean that the rule for assignment, or rather the `(H,E)[x <- v]` notation, is wrong as it doesn't handle the uninitalized case correctly?) I don't like this, I would prefer `mut x` to allocate rather than the first write, because it means the semantics is closer to the syntax -- and it makes the performance of assignment easier to reason about.